### PR TITLE
fix: Decoupled proactive token refresh from FirebaseApp

### PR DESCRIPTION
**Pull request by hiranya911**
Our existing code for OAuth2 token management is quite complicated. This is due to `FirebaseApp` doing too much -- namely handling token caching as well proactive token refreshing. The proactive token refresh was added to make our RTDB connections more stable, but having this logic in `FirebaseApp` impacts other APIs in the SDK too. For instance this is why the SDK won't let the Node.js process terminate, unless you call `app.delete()`. 

In order to simplify our token management, and align more with other GCP libraries that deal with tokens, I'm proposing to separate the proactive token refresh from the `FirebaseApp` class. To that end this PR implements the following changes:

1. Remove the proactive token refresh logic from `FirebaseApp`.
2. Simplify the token caching and fetching behavior.
3. Treat any cached token that's at most 5 minutes away from expiration to be already expired. 
4. Implement the proactive token refresh as part of the `database` service, where it will only get initialized if a developer starts an RTDB connection.
5. Remove the existing complicated retry logic in the proactive token refresh.

Change no. 3 is pretty standard across GCP libraries (see [here](https://github.com/googleapis/google-auth-library-nodejs/blob/fbad5d383be691099a7dab29dca27d5263c4dfb1/src/auth/oauth2client.ts#L1320-L1325) and [here](https://github.com/googleapis/google-auth-library-nodejs/blob/fbad5d383be691099a7dab29dca27d5263c4dfb1/src/auth/authclient.ts#L94)). We never implemented this in the Admin Node.js SDK since the token refresher was indirectly handling it. But it appears in certain environments we cannot rely on the token refresher to keep the auth token current. Therefore by eagerly treating cached tokens as expired, we can mitigate issues like #1193 entirely.

As for change no. 5, the existing retry logic in the token refresher is rather error prone and redundant. If something goes wrong with a proactive refresh attempt, the worst that could happen is any active RTDB connections dropping out upon the expiration of the current token. The RTDB SDK automatically reconnects when this happens (provided it can fetch a new token at that point). This is what we do in our Java SDK too, so there's some prior art as well.

### CODE REVIEW

File: src/database/database-internal.ts [hunk 1]
```diff
@@ -96,9 +106,45 @@ export class DatabaseService {
...
+    if (!this.tokenRefresh) {
+      this.tokenRefresh = true;
+      this.appInternal.INTERNAL.addAuthTokenListener(this.onTokenChange);
+    }
+
     return db;
   }
 
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  private onTokenChange(_: string): void {
```
Comment (inlined) [inline src/database/database-internal.ts:119]: IIRC there was no other token refresh listener. If the right action is to immediately ask for a new token from app.INTERNAL should we just remove the parameter?

File: src/database/database-internal.ts [hunk 2]
```diff
@@ -96,9 +106,45 @@ export class DatabaseService {
...
+        const delayMillis = token.expirationTime - TOKEN_REFRESH_THRESHOLD_MILLIS - Date.now();
+        // If the new token is set to expire soon (unlikely), do nothing. Somebody will eventually
+        // notice and refresh the token, at which point this callback will fire again.
+        if (delayMillis > 0) {
+          this.scheduleTokenRefresh(delayMillis);
+        }
+      })
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      .catch((_) => {
+        // Unlikely error, since a new token was just fetched before the callback fired.
```
Comment (inlined) [inline src/database/database-internal.ts:131]: I'm used to the idea that we should at least log a caught error. Dropping it silently sounds like we'll have a really hard time debugging any bugs here.

File: src/database/database-internal.ts [hunk 3]
```diff
@@ -96,9 +106,45 @@ export class DatabaseService {
...
+        // Just here to prevent the promise chain from crashing.
+      });
+  }
+
+  private scheduleTokenRefresh(delayMillis: number): void {
+    clearTimeout(this.tokenRefreshTimeout);
+    this.tokenRefreshTimeout = setTimeout(() => {
+      this.appInternal.INTERNAL.getToken(true)
```
Comment (inlined) [inline src/database/database-internal.ts:139]: nit: I don't know if it conflicts with the style guide here, but style guides internally say you should comment to name any boolean variable names:

```js
this.appInternal.INTERNAL.getToken(/*forceRefresh=*/true);
```

File: src/firebase-app.ts [hunk 4]
```diff
@@ -57,129 +59,80 @@ export interface FirebaseAccessToken {
...
+  public getToken(forceRefresh = false): Promise<FirebaseAccessToken> {
+    if (forceRefresh || this.shouldRefresh()) {
+      return this.refreshToken();
     }
+
+    return Promise.resolve(this.cachedToken_);
+  }
+
+  private refreshToken(): Promise<FirebaseAccessToken> {
+    return Promise.resolve(this.credential_.getAccessToken())
```
Comment (inlined) [inline src/firebase-app.ts:78]: I'm confused. isn't getAccesssToken() already returning a promise? Why do we need to resolve the promise instead of just chaining?

**Review by inlined — APPROVED** [review]
Absolutely awesome turnaround time. Thanks Hiranya!

File: src/database/database-internal.ts (hunk 1 above)
Comment (hiranya911) [reply src/database/database-internal.ts:119]: I think an even better approach would be to change the signature of the listener to accept a `FirebaseAccessToken` which contains both token and expiry time. That way we wouldn't have to call `getToken()` in the listener at all (we currently do it to get hold of the expiry time). 

However, this API currently mirrors an API in the JS SDK:

https://github.com/firebase/firebase-js-sdk/blob/ec6ccc7d319732588ceacae0824d6c7551afb026/packages/app-types/private.d.ts#L76

The RTDB implementation expects our App and App.INTERNAL APIs to be compatible with the JS SDK. Therefore we'll need to check with the JS SDK team before making any changes to the listener signature.

File: src/firebase-app.ts (hunk 4 above)
Comment (hiranya911) [reply src/firebase-app.ts:78]: I thought so too :)

But when I made the change, it actually caused a test failure here:
https://github.com/firebase/firebase-admin-node/blob/994fd43e36c0824deba176763010dc141e9bb0f4/test/unit/firebase-app.spec.ts#L677-L691

This is because the `credential` in this case returns a value which cannot be chained. I'm not sure how much we care about this sort of edge cases, but for now I've opted to keep it as is.

File: src/database/database-internal.ts (hunk 2 above)
Comment (hiranya911) [reply src/database/database-internal.ts]: Done

File: src/database/database-internal.ts (hunk 3 above)
Comment (hiranya911) [reply src/database/database-internal.ts]: Done

File: src/database/database-internal.ts [hunk 5]
```diff
@@ -28,9 +28,13 @@ import { getSdkVersion } from '../utils/index';
 
 import Database = database.Database;
 
+const TOKEN_REFRESH_THRESHOLD_MILLIS = 5 * 60 * 1000;
+
 export class DatabaseService {
 
   private readonly appInternal: FirebaseApp;
+  private tokenRefresh: boolean;
```
Comment (schmidt-sebastian) [inline src/database/database-internal.ts:36]: This doesn't sound like a boolean variable to me (but naming is hard and criticizing is easy)

File: src/database/database-internal.ts [hunk 6]
```diff
@@ -96,9 +106,43 @@ export class DatabaseService {
...
+      .then((token) => {
+        const delayMillis = token.expirationTime - TOKEN_REFRESH_THRESHOLD_MILLIS - Date.now();
+        // If the new token is set to expire soon (unlikely), do nothing. Somebody will eventually
+        // notice and refresh the token, at which point this callback will fire again.
+        if (delayMillis > 0) {
+          this.scheduleTokenRefresh(delayMillis);
+        }
+      })
+      .catch((err) => {
+        console.error('Unexpected error while attempting to schedule a token refresh:', err);
```
Comment (schmidt-sebastian) [inline src/database/database-internal.ts:130]: It think the error most likely occurred during `getToken()` and not during the scheduling step.

File: src/database/database-internal.ts [hunk 7]
```diff
@@ -96,9 +106,43 @@ export class DatabaseService {
 
       this.databases[dbUrl] = db;
     }
+
+    if (!this.tokenRefresh) {
+      this.tokenRefresh = true;
```
Comment (schmidt-sebastian) [inline src/database/database-internal.ts:111]: Optional suggestion:

You can make this a little simpler if you combine `tokenRefresh` and `tokenRefreshTimeout`. You could initialize `tokenRefreshTimeout` to `undefined` initially and then set the schedule right here, which seems more intuitive to me. With the current setup, I have to look at `onTokenChange` to figure out the scheduling.

I also wonder if we should use `setInterval`. It seems a little more appropriate, but might make things more complicated in some places.

File: src/firebase-app.ts [hunk 8]
```diff
@@ -57,129 +59,80 @@ export interface FirebaseAccessToken {
...
-            'present at https://console.firebase.google.com/iam-admin/serviceaccounts/project. If ' +
-            'not, generate a new key file at ' +
-            'https://console.firebase.google.com/project/_/settings/serviceaccounts/adminsdk.';
-          }
-
-          throw new FirebaseAppError(AppErrorCodes.INVALID_CREDENTIAL, errorMessage);
-        });
-
-      return this.cachedTokenPromise_;
+  public getToken(forceRefresh = false): Promise<FirebaseAccessToken> {
```
Comment (schmidt-sebastian) [inline src/firebase-app.ts:69]: Can the RTDB component just set `forceRefresh` to true if the expiration time has been hit? That way, we only need to deal with TOKEN_REFRESH_THRESHOLD_MILLIS in one place.

**Review by schmidt-sebastian — APPROVED** [review]

File: src/database/database-internal.ts (hunk 5 above)
Comment (lahirumaramba) [reply src/database/database-internal.ts:36]: `refreshToken` or `shouldRefreshToken` might work... I agree with "naming is hard and criticizing is easy" :)

File: src/database/database-internal.ts [hunk 9]
```diff
@@ -50,6 +54,12 @@ export class DatabaseService {
    * @internal
    */
   public delete(): Promise<void> {
+    if (this.tokenRefresh) {
+      this.appInternal.INTERNAL.removeAuthTokenListener(this.onTokenChange);
+      clearTimeout(this.tokenRefreshTimeout);
```
Comment (lahirumaramba) [inline src/database/database-internal.ts:59]: Does it matter the what order we disarm here? Could this cause a race condition?

**Review by lahirumaramba — APPROVED** [review]
Thanks, Hiranya! LGTM!
Left a couple comments.

File: src/database/database-internal.ts (hunk 6 above)
Comment (hiranya911) [reply src/database/database-internal.ts:130]: True. But overall this function is attempting to schedule a new token refresh event. It calls `getToken()` just to figure out when to schedule that event.

File: src/database/database-internal.ts (hunk 9 above)
Comment (hiranya911) [reply src/database/database-internal.ts:59]: I felt like we should unregister the listener before clearing out the timer. If we clear the timer first, theoretically it's possible for the listener to fire again and re-initialize the timer.

File: src/database/database-internal.ts (hunk 7 above)
Comment (hiranya911) [reply src/database/database-internal.ts]: This code is just intended to register the token change listener. There's no guarantee that the listener will fire immediately. So theoretically it is possible for `tokenRefresh` to be true, while `tokenRefreshTimeout` is not yet set. What we really need is a way to track whether we have registered the listener or not (so we can avoid duplicate registrations, and also unregister it at delete). So I'm renaming `tokenRefresh` to `tokenListenerRegistered`. That should clear up any confusions, and also address the above comment regarding naming.

As for `setInterval()`, given that we can receive tokens with arbitrary expiration times from auth (we've seen tokens with expiry times like 30min and 45min), an interval scheduler is not very useful. So we use the token listener to watch for token changes, and schedule refresh events according to the expiry time we get.

File: src/firebase-app.ts (hunk 8 above)
Comment (hiranya911) [reply src/firebase-app.ts:69]: These 2 constants are used for slightly different purposes:

1. In firebase-app it is used to determine how early we should treat a valid token as expired.
2. In database-internal it is used to determine when to schedule the next refresh event.

These don't have to be the same value. For instance, we can have FirebaseApp invalidate a token 5 minutes before it expires, but have RTDB trigger the token refresh 10 minutes before the current token expires.

To make the above distinction clear, I'm renaming the constant in firebase-app to `TOKEN_EXPIRY_THRESHOLD_MILLIS`. This constant doesn't have anything to do with refreshing tokens. It's simply when we treat the token as expired.

File: src/database/database-internal.ts (hunk 5 above)
Comment (hiranya911) [reply src/database/database-internal.ts]: Renamed to `tokenListenerRegistered`. See below.
